xdsl/dialects/hw.py

```diff
@@ -112,6 +121,142 @@ def print_parameters(self, printer: Printer) -> None:
+        self.symbol_table: Mapping[StringAttr, InnerSymTarget] = dict()
+        super().__init__()
+
+    def verify(self, op: Operation):
+        # Insist that ops with InnerSymbolTable's provide a Symbol, this is
+        # essential to how InnerRef's work.
+        if not op.has_trait(trait := SymbolOpInterface):
+            raise VerifyException(
+                f"Operation {op.name} must have trait {trait.__name__}"
+            )
+
+        # InnerSymbolTable's must be directly nested within an InnerRefNamespace,
+        # however don’t test InnerRefNamespace’s symbol lookups
+        parent = op.parent_op()
+        if (
+            parent is None
+            or len(parent.get_traits_of_type(trait := InnerRefNamespace)) != 1
+        ):
+            raise VerifyException(
+                f"Operation {op.name} with trait {type(self).__name__} must have a parent with trait {trait.__name__}"
+            )
+
+
+class InnerSymbolTableCollection:
+    """This class represents an InnerSymbolTable collection."""
+
+    def __init__(self, inner_ref_ns_op: Operation | None = None) -> None:
+        self.symbol_tables: dict[Operation, InnerSymbolTable] = {}
+        if inner_ref_ns_op is not None:
+            self.populate_and_verify_tables(inner_ref_ns_op)
+
+    def get_inner_symbol_table(self, op: Operation) -> InnerSymbolTable:
+        """Returns the InnerSymolTable trait, ensuring `op` is in the collection"""
+        if not op.has_trait(trait := InnerSymbolTable):
+            raise VerifyException(
+                f"Operation {op.name} should have {trait.__name__} trait"
+            )
+        if op not in self.symbol_tables:
+            self.symbol_tables[op] = InnerSymbolTable(op)
+        return self.symbol_tables[op]
+
+    def populate_and_verify_tables(self, inner_ref_ns_op: Operation):
+        """Populate tables for all InnerSymbolTable operations in the given InnerRefNamespace operation, verifying each."""
+        # Gather top-level operations that have the InnerSymbolTable trait.
+        inner_sym_table_ops = (
+            op for op in inner_ref_ns_op.walk() if op.has_trait(InnerSymbolTable)
+        )
+
+        # Construct the tables
+        for op in inner_sym_table_ops:
+            if op in self.symbol_tables:
+                raise VerifyException(
+                    f"Trying to insert the same op {op} twice in symbol tables"
+                )
+            self.symbol_tables[op] = InnerSymbolTable(op)
+
+
+class InnerRefUserOpInterface(OpTrait):
+    """This interface describes an operation that may use a `InnerRef`. This
+    interface allows for users of inner symbols to hook into verification and
+    other inner symbol related utilities that are either costly or otherwise
+    disallowed within a traditional operation."""
+
+    def verify_inner_refs(self, op: Operation, namespace: "InnerRefNamespace"):
+        """Verify the inner ref uses held by this operation."""
+        ...
```

Comment:
Is this abstract?

Reply:
Just virtual AFAICT. It allows users of inner symbols a way to hook into verifications but there’s no obligation to actually do anything.